File: src/services/profile_generation_service.py
from __future__ import annotations

from collections import Counter

from src.models.domain import CandidateProfile, CompanyDossier, EvidenceGap, ExperienceRecord
from src.services.hiring_judgment_service import HiringJudgmentService


class ProfileGenerationService:
    def __init__(self, hiring_judgment_service: HiringJudgmentService):
        self.hiring_judgment_service = hiring_judgment_service

    def generate(
        self,
        experiences: list[ExperienceRecord],
        dossiers: list[CompanyDossier],
        gaps_by_experience_id: dict[int, list[EvidenceGap]],
    ) -> CandidateProfile:
        themes = self._extract_themes(experiences)
        weak_spots = self._extract_weak_spots(gaps_by_experience_id)
        positioning = self.hiring_judgment_service.evaluate_positioning(experiences)
        companies = ", ".join(experience.company for experience in experiences[:3])
        return CandidateProfile(
            career_arc=(
                f"当前主线经历集中在 {companies}，整体更像一个能把复杂项目往前推的产品型候选人，"
                "但竞争力高低取决于关键案例里的证据密度。"
            ),
            strongest_themes=themes,
            weak_spots=weak_spots,
            positioning_boundary=(
                "适合把证据最强的方向作为主卖点，不建议在没有结果和 ownership 证据的情况下，"
                "直接把自己包装成全能型候选人。"
            ),
            recommended_main_lane=positioning["recommended_main_lane"],
            conservative_target_strategy=positioning["conservative_target_strategy"],
        )

    def _extract_themes(self, experiences: list[ExperienceRecord]) -> list[str]:
        counter: Counter[str] = Counter()
        joined = " ".join(
            " ".join(experience.projects + experience.responsibilities + experience.outcomes).lower()
            for experience in experiences
        )
        theme_rules = {
            "AI / Agent / workflow": ["ai", "agent", "workflow", "llm", "automation"],
            "Growth / retention": ["growth", "retention", "留存", "转化", "用户增长"],
            "Cross-functional execution": ["跨团队", "stakeholder", "alignment", "协同", "推动"],
            "Data-driven product work": ["sql", "data", "metric", "roi", "ab test", "实验"],
        }
        for theme, keywords in theme_rules.items():
            counter[theme] = sum(keyword in joined for keyword in keywords)
        ranked = [theme for theme, score in counter.most_common() if score > 0]
        return ranked[:3] or ["Complex product execution"]

    def _extract_weak_spots(self, gaps_by_experience_id: dict[int, list[EvidenceGap]]) -> list[str]:
        counter: Counter[str] = Counter()
        labels = {
            "result": "结果证据偏弱",
            "ownership": "ownership 叙事不够扎实",
            "scope": "项目范围和业务上下文不够清楚",
            "decision": "关键判断与取舍不够具体",
            "tradeoff": "tradeoff 讲法还不成熟",
            "failure": "失败与调整案例偏少",
            "influence": "影响他人协作的证据不足",
        }
        for gaps in gaps_by_experience_id.values():
            for gap in gaps:
                counter[labels[gap.gap_type]] += 1
        return [label for label, _ in counter.most_common(3)] or ["表达基础还行，但需要更多结果和决策证据"]

File: apps/api/src/services/profile-generation-service.ts
import type { CandidateProfile, CompanyDossier, EvidenceGap, ExperienceRecord, GoalSetupState } from "../domain.js";
import {
  hasConfirmedStrongOwnership,
  hasOwnershipRestriction,
  type ClaimRestrictionsByExperienceId,
} from "./claim-safety.js";
import { HiringJudgmentService } from "./hiring-judgment-service.js";

export class ProfileGenerationService {
  constructor(private readonly hiringJudgmentService: HiringJudgmentService) {}

  generate(
    experiences: ExperienceRecord[],
    _dossiers: CompanyDossier[],
    gapsByExperienceId: Record<number, EvidenceGap[]>,
    goalSetup?: GoalSetupState | null,
    restrictionsByExperienceId: ClaimRestrictionsByExperienceId = {},
  ): CandidateProfile {
    const strongestThemes = this.extractThemes(experiences);
    const weakSpots = this.extractWeakSpots(gapsByExperienceId);
    const positioning = this.hiringJudgmentService.evaluatePositioning(experiences, goalSetup);
    const companies = experiences.slice(0, 3).map((experience) => experience.company).join(", ");
    const targetRoleHint = goalSetup?.targetRole.trim()
      ? `这次求职先按「${goalSetup.targetRole.trim()}」来梳理，`
      : "";
    const sellingPointHint = goalSetup?.mainSellingPoint.trim()
      ? `建议重点突出「${goalSetup.mainSellingPoint.trim()}」。`
      : "建议重点突出事实依据最充分的能力。";
    const claimRestrictions = Object.values(restrictionsByExperienceId)
      .flat()
      .filter((restriction, index, all) =>
        all.findIndex((candidate) => candidate.code === restriction.code) === index,
      );
    const hasSafeStrongOwnership = experiences.some((experience) =>
      hasConfirmedStrongOwnership(experience)
      && !hasOwnershipRestriction(restrictionsByExperienceId[experience.id] ?? []),
    );
    const careerEvidence = hasSafeStrongOwnership
      ? "现有经历能说明你有推动复杂项目落地的能力"
      : "现有材料能确认你参与过相关项目，但个人贡献暂时不宜写得过重";
    const baseBoundary = goalSetup?.doNotOversell.trim()
      ? `这次明确不要夸大「${goalSetup.doNotOversell.trim()}」。没有明确结果或个人贡献依据的内容，也不宜写成你的核心优势。`
      : "适合重点突出事实依据最充分的方向；没有明确结果或个人贡献依据的内容，不宜写成核心优势。";
    const restrictionBoundary = claimRestrictions.length > 0
      ? ` 写作时还需注意：${claimRestrictions.map((restriction) => restriction.description).join("；")}`
      : "";
    return {
      careerArc: `${targetRoleHint}主要经历集中在 ${companies}，${careerEvidence}。关键案例中的事实越具体，表达越有说服力。${sellingPointHint}`,
      strongestThemes,
      weakSpots,
      positioningBoundary: `${baseBoundary}${restrictionBoundary}`,
      recommendedMainLane: positioning.recommendedMainLane,
      conservativeTargetStrategy: positioning.conservativeTargetStrategy,
    };
  }

  private extractThemes(experiences: ExperienceRecord[]): string[] {
    const joined = experiences.flatMap((experience) => [...experience.projects, ...experience.responsibilities, ...experience.outcomes]).join(" ").toLowerCase();
    const rules = {
      "AI、智能体与工作流产品": ["ai", "agent", "workflow", "llm", "automation"],
      "增长与留存": ["growth", "retention", "留存", "转化", "用户增长"],
      "跨团队协作与推进": ["跨团队", "stakeholder", "alignment", "协同", "推动"],
      "数据驱动的产品工作": ["sql", "data", "metric", "roi", "ab test", "实验"],
    };
    return Object.entries(rules)
      .map(([theme, keywords]) => ({
        theme,
        score: keywords.filter((keyword) => joined.includes(keyword.toLowerCase())).length,
      }))
      .filter((item) => item.score > 0)
      .sort((left, right) => right.score - left.score)
      .slice(0, 3)
      .map((item) => item.theme);
  }

  private extractWeakSpots(gapsByExperienceId: Record<number, EvidenceGap[]>): string[] {
    const labels: Record<string, string> = {
      result: "实际结果还不够明确",
      ownership: "个人贡献还不够清楚",
      scope: "项目背景和影响范围还不够清楚",
      decision: "关键判断与取舍不够具体",
      tradeoff: "关键取舍还不够具体",
      failure: "问题与调整过程较少",
      influence: "协作推进的事实依据不足",
    };
    const counts = new Map<string, number>();
    for (const gaps of Object.values(gapsByExperienceId)) {
      for (const gap of gaps) {
        const label = labels[gap.gapType];
        counts.set(label, (counts.get(label) ?? 0) + 1);
      }
    }
    return [...counts.entries()].sort((a, b) => b[1] - a[1]).slice(0, 3).map(([label]) => label);
  }
}
